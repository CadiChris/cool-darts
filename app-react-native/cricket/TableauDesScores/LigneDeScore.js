import React from 'react'
import { connect } from 'react-redux'
import { Text, View, TouchableHighlight } from 'react-native'
import { Col, Row } from "react-native-easy-grid"
import { Styles, Tailles } from './Styles'
import { lancerFlechette } from './TableauDesScores.actions'

class LigneDeScore extends React.Component {
  constructor(props) {
    super(props)
  }

  render() {
    const {score} = this.props
    return (
      <Row style={Styles.bordureBasse}>
        <Col style={{justifyContent: 'center'}} size={Tailles.largeurDeLaColonneJoueur}>
          <Text style={{paddingLeft: 10}}>{score.joueur} - {score.points}</Text>
        </Col>

        {
          Object.keys(score.cible).map((chiffre, index) => {
            const {touches, ferme} = score.cible[chiffre]
            return (
              <Col style={[{flex:1, flexDirection:'row', alignItems:'stretch'}, Styles.bordureDroite, ferme ? Styles.chiffreFerme : {}]} key={index}>
                <TouchableHighlight onPress={() => this.lancerSimpleDans(chiffre)} style={{alignSelf: 'stretch'}}>
                  <View>
                    <Text> </Text>
                    <Text style={{textAlign: 'center'}}>{'X'.repeat(touches)}</Text>
                  </View>
                </TouchableHighlight>
              </Col>
            )
          })
        }
      </Row>
    )
  }

  lancerSimpleDans(chiffre) {
    const {dispatch, score} = this.props
    dispatch(lancerFlechette(score.joueur, chiffre, 1))
  }
}

export default connect()(LigneDeScore)