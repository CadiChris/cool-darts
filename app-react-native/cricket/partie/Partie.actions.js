export function inscrireJoueur(joueur) {
  return {
    type: 'INSCRIRE_JOUEUR',
    joueur
  }
}

export function demarrerPartie() {
  return {type: 'DEMARRER_PARTIE'}
}