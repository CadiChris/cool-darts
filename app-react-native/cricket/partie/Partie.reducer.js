const LANCEUR = {
  nom: '',
  flechettesRestantes: 3,
  dernierLancer :{
    chiffre: null, touches: 0
  }
}

const stateInitial = {
  phase: 'INSCRIPTION',
  lanceur: LANCEUR,
  joueurs: [],
  peutDemarrer: false
}

export default function partie(state = stateInitial, action) {
  switch (action.type) {
    case 'INSCRIRE_JOUEUR':
      return {
        ...state,
        joueurs: [...state.joueurs, action.joueur],
        peutDemarrer: state.joueurs.length > 0
      }

    case 'DEMARRER_PARTIE':
      return {
        ...state,
        phase: 'EN_COURS',
        lanceur: {
          ...state.lanceur,
          nom: state.joueurs[0]
        }
      }

    case 'LANCER_FLECHETTE' :
      return {
        ...state,
        lanceur: state.lanceur.flechettesRestantes > 1
          ? {
            ...state.lanceur,
            flechettesRestantes: state.lanceur.flechettesRestantes - 1
          } : {
            ...LANCEUR,
            nom: state.joueurs[(state.joueurs.indexOf(state.lanceur.nom) + 1) % state.joueurs.length],
          }
      }

    case 'CHIFFRE' :
      const aDeja3 = state.lanceur.dernierLancer.touches === 3
      return {
        ...state,
        lanceur: {
          ...state.lanceur,
          dernierLancer: {
            chiffre: action.chiffre,
            touches: aDeja3 ? 1 : state.lanceur.dernierLancer.touches + 1
          }
        }
      }

    default:
      return state
  }
}