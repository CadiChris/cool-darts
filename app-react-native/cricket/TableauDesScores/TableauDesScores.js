import React from 'react'
import { connect } from 'react-redux'
import { StyleSheet, Text, View } from 'react-native'
import { Grid } from "react-native-easy-grid"
import EnTete from './EnTete'
import LigneDeScore from './LigneDeScore'


class TableauDesScores extends React.Component {
  constructor(props) {
    super(props)
  }

  render() {
    const {scores} = this.props
    return (
      <View style={{flex: 1}}>
        <Grid style={{flexDirection: 'column'}}>
          <EnTete />
          {
            scores.map((score, index) => <LigneDeScore score={score} key={index}/>)
          }
        </Grid>
      </View>
    )
  }
}

function mapStateToProps(state) {
  return {
    scores: [...state.tableauDesScores.scores]
  }
}

export default connect(mapStateToProps)(TableauDesScores)
