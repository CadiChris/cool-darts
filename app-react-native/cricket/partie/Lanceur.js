import React from 'react'
import { StyleSheet, Text, View, Button } from 'react-native'
import { connect } from 'react-redux'
import { lancerFlechette } from '../tableau-des-scores/TableauDesScores.actions'

class Lanceur extends React.Component {

  constructor(props) {
    super(props)
    this.state = {
      ...this.aucuneSaisie()
    }
  }

  render() {
    const chiffres_radio = [15, 16, 17, 18, 19, 20].map(c => {
      return {label: c, value: c}
    })
    const touches_radio = [
      {label: 'Simple', value: 1},
      {label: 'Double', value: 2},
      {label: 'Triple', value: 3},
    ]

    const {lanceur, chiffre, touches} = this.props
    const peutLancer = this.laSaisieEstValide()

    return (
      <View>
        <Text>{lanceur.nom} - {'O'.repeat(lanceur.flechettesRestantes)}</Text>
        <Text>{chiffre} - {touches ? touches_radio[touches - 1].label : ''}</Text>
        <View style={{flexDirection: 'row'}}>
          <Button title="Lancer" onPress={() => this.lancer()} disabled={!peutLancer}/>
          <Button title="ManquÃ©" onPress={() => this.manquer()}/>
        </View>
      </View>
    )
  }

  aucuneSaisie() {
    return {
      chiffre: null,
      touches: null
    }
  }

  noterChiffre(chiffre) {
    this.setState({
      ...this.state,
      chiffre
    })
  }

  noterTouches(touches) {
    this.setState({
      ...this.state,
      touches
    })
  }

  laSaisieEstValide() {
    return this.props.chiffre !== null && this.props.touches !== null
  }

  lancer() {
    const {dispatch, lanceur, chiffre, touches} = this.props
    dispatch(lancerFlechette(lanceur.nom, chiffre, touches))
    this.setState(this.aucuneSaisie())
  }

  manquer() {
    const {dispatch, lanceur} = this.props
    dispatch(lancerFlechette(lanceur.nom, 0, 1))
  }
}

function mapStateToProps(state) {
  return {
    chiffre: state.partie.lanceur.dernierLancer.chiffre,
    touches: state.partie.lanceur.dernierLancer.touches
  }
}
export default connect(mapStateToProps)(Lanceur)