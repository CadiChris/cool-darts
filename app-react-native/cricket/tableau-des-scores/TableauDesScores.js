import React from 'react'
import { StyleSheet, Text, View } from 'react-native'
import EnTete from './EnTete'
import LigneDeScore from './LigneDeScore'

import { Grid, Row } from "react-native-easy-grid";

export default () => {
  return (
    <View style={{flex:1, alignItems:'stretch'}}>
      <Grid>
        <Row style={{height: 45}}><EnTete /></Row>
        <Row><LigneDeScore /></Row>
      </Grid>
    </View>
  )
}