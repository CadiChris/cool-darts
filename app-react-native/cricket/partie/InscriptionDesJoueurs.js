import React from 'react'
import { connect } from 'react-redux'
import { View, Button, TextInput } from 'react-native'
import { inscrireJoueur, demarrerPartie } from './Partie.actions'

class InscriptionDesJoueurs extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      joueur: ''
    }
  }

  render() {
    const {peutDemarrer} = this.props
    const {joueur} = this.state
    return (
      <View style={{height: 100}}>
        <View style={{flex: 1, flexDirection: 'row'}}>
          <TextInput
            value={joueur}
            onChangeText={(text) => this.setState({joueur: text})}
            style={{height: 40, width: 100}}
            placeholder="Joueur"/>
          <Button title="Nouveau joueur" onPress={() => this.ajouterJoueur()}/>
        </View>
        {
          peutDemarrer ? <Button title="GO" onPress={() => this.demarrerLaPartie()}/> : null
        }
      </View>
    )
  }

  ajouterJoueur() {
    const {dispatch} = this.props
    const {joueur}=this.state
    dispatch(inscrireJoueur(joueur))
    this.setState({joueur: ''})
  }

  demarrerLaPartie() {
    const {dispatch} = this.props
    dispatch(demarrerPartie())
  }
}

function mapStateToProps(state) {
  return {
    peutDemarrer: state.partie.peutDemarrer
  }
}

export default connect(mapStateToProps)(InscriptionDesJoueurs)