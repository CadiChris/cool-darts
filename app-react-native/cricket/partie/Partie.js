import React from 'react'
import { connect } from 'react-redux'
import { View, Text } from 'react-native'
import Lanceur from './Lanceur'
import InscriptionDesJoueurs from './InscriptionDesJoueurs'

class Partie extends React.Component {
  constructor(props) {
    super(props)
  }

  render() {
    const {peutDemarrer, phase} = this.props
    console.log(phase)
    return (
      <View>
        {
          phase === 'INSCRIPTION'
            ? <InscriptionDesJoueurs />
            : <Lanceur lanceur={{nom: 'Christophe'}}/>

        }
        <Text>{peutDemarrer ? 'READY' : 'Pas encore'}</Text>
      </View>
    )
  }
}

function mapStateToProps(state) {
  console.log(state)
  return {
    peutDemarrer: state.partie.peutDemarrer,
    phase: state.partie.phase
  }
}

export default connect(mapStateToProps)(Partie)