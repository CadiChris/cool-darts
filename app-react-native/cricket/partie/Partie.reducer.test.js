import partie from './Partie.reducer'
import { inscrireJoueur, demarrerPartie } from './Partie.actions'
import { lancerFlechette, chiffre } from '../tableau-des-scores/TableauDesScores.actions'

it('retourne le state initial', () => {
  expect(partie(undefined, {})).toMatchSnapshot()
})

it('inscrit des joueurs', () => {
  const partieAvec1Joueur = partie(partieDeTest(), inscrireJoueur('J1'))
  expect(partieAvec1Joueur).toMatchSnapshot()

  const partieAvec2Joueurs = partie(partieAvec1Joueur, inscrireJoueur('J2'))
  expect(partieAvec2Joueurs).toMatchSnapshot()
})

it('démarre la partie', () => {
  const partieAvec1Joueur = partie(partieDeTest(), inscrireJoueur('J1'))

  expect(partie(partieAvec1Joueur, demarrerPartie())).toMatchSnapshot()
})

it('compte les fléchettes restantes', () => {
  const state = {
    lanceur: {
      nom: 'J1',
      flechettesRestantes: 3
    }
  }
  expect(partie(state, lancerFlechette('', 0, 0))).toEqual({lanceur: {nom: 'J1', flechettesRestantes: 2}})
})

it('note la dernière fléchette lancée', () => {
  const state = {
    lanceur: {
      dernierLancer: {
        chiffre: null, touches: 0
      }
    }
  }

  const expectLesTouches = (lancer, touchesAttendues) => {
    expect(lancer.lanceur.dernierLancer).toEqual({chiffre: 20, touches: touchesAttendues})
  }

  const jet1 = partie(state, chiffre(20))
  const jet2 = partie(jet1, chiffre(20))
  const jet3 = partie(jet2, chiffre(20))
  const jet4 = partie(jet3, chiffre(20))

  expectLesTouches(jet1, 1)
  expectLesTouches(jet2, 2)
  expectLesTouches(jet3, 3)
  expectLesTouches(jet4, 1)
})

it('change de lanceur au dernier lancer', () => {
  const stateDuPremierLanceur = {
    joueurs: ['J1', 'J2'],
    lanceur: {
      nom: 'J1',
      flechettesRestantes: 1,
      dernierLancer: {
        chiffre: null,
        touches: 0
      }
    }
  }
  const stateDuSecondLanceur = {
    joueurs: ['J1', 'J2'],
    lanceur: {
      nom: 'J2',
      flechettesRestantes: 3,
      dernierLancer: {
        chiffre: null,
        touches: 0
      }
    }
  }
  expect(partie(stateDuPremierLanceur, lancerFlechette('', 0, 0))).toEqual(stateDuSecondLanceur)

  const stateDuSecondLanceurDerniereFlechette = {
    joueurs: ['J1', 'J2'],
    lanceur: {
      nom: 'J2',
      flechettesRestantes: 1
    }
  }
  expect(partie(stateDuSecondLanceurDerniereFlechette, lancerFlechette('', 0, 0)).lanceur.nom).toEqual('J1')
})

function partieDeTest() {
  return partie(undefined, {})
}