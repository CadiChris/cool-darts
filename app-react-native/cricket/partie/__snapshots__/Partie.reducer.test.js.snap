// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`d√©marre la partie 1`] = `
Object {
  "joueurs": Array [
    "J1",
  ],
  "lanceur": Object {
    "dernierLancer": Object {
      "chiffre": null,
      "touches": 0,
    },
    "flechettesRestantes": 3,
    "nom": "J1",
  },
  "peutDemarrer": false,
  "phase": "EN_COURS",
}
`;

exports[`inscrit des joueurs 1`] = `
Object {
  "joueurs": Array [
    "J1",
  ],
  "lanceur": Object {
    "dernierLancer": Object {
      "chiffre": null,
      "touches": 0,
    },
    "flechettesRestantes": 3,
    "nom": "",
  },
  "peutDemarrer": false,
  "phase": "INSCRIPTION",
}
`;

exports[`inscrit des joueurs 2`] = `
Object {
  "joueurs": Array [
    "J1",
    "J2",
  ],
  "lanceur": Object {
    "dernierLancer": Object {
      "chiffre": null,
      "touches": 0,
    },
    "flechettesRestantes": 3,
    "nom": "",
  },
  "peutDemarrer": true,
  "phase": "INSCRIPTION",
}
`;

exports[`retourne le state initial 1`] = `
Object {
  "joueurs": Array [],
  "lanceur": Object {
    "dernierLancer": Object {
      "chiffre": null,
      "touches": 0,
    },
    "flechettesRestantes": 3,
    "nom": "",
  },
  "peutDemarrer": false,
  "phase": "INSCRIPTION",
}
`;
